tools/meson-vcs-tag.sh

```diff
@@ -29,5 +29,8 @@ else
         fi
     fi
     [ -z "$c" ] && c="${fallback}"
-    echo "$c" | sed 's/^v//; s/-rc/~rc/'
+    # Replace any hyphens with carets which are allowed in versions by pacman whereas hyphens are not. Git
+    # versions with carets will also sort higher than their non-git version counterpart both in pacman
+    # versioning and in version comparision spec versioning.
+    echo "$c" | sed 's/^v//; s/-/^/g'
```

Comment:
i don't get this, can you explain what this is about?

Reply:
Currently this generates something like `255-abcde-fgh` which the `pacman` package manager does not like at all. I'm proposing we replace this with `255^abcde^fgh` which `pacman` does like and is still guaranteed to sort higher than `255`.

Reply:
never seen commit strings like that. does any other project use that? won#t this confuse git, because git uses `^` for its own purposes?

Reply:
Why would this confuse git? This never gets seen by git, all meson-vcs-tag.sh does is generate versions for use with our version-tag option or packaging specs if the version tag is not set explicitly by the user.

Reply:
hmm, i mean, don#t downstreams take our upstream packages and then turn them into tags on their own?

Reply:
@poettering Yes but I'm looking at using the downstream packaging specs in our mkosi CI, which means I want them to use versions generated by meson-vcs-tag.sh.

Reply:
Both debs and rpms allow versioning with `^` and `~`. `^` were added more or less for this purpose.

E.g. see https://docs.fedoraproject.org/en-US/packaging-guidelines/Versioning/#_handling_non_sorting_versions_with_tilde_dot_and_caret:
> The caret symbol (‘^’) is used before a version component which must sort later than any non-caret component. It is used for post-release snapshots

Reply:
But do we actually need this? With the switch to `~`, we will only use `[v0-9.~]`. Where would the dash originate?

Reply:
@keszybz 

```
➜  systemd git:(main) tools/meson-vcs-tag.sh "." error
255-1734-g23f9af6
➜  systemd git:(main) git checkout rc-versioning
Switched to branch 'rc-versioning'
Your branch is up to date with 'fork/rc-versioning'.
➜  systemd git:(rc-versioning) tools/meson-vcs-tag.sh "." error
255^1735^g42d77d0
```